components/Banner/src/MDCBannerView.m

```diff
@@ -0,0 +1,474 @@
+    NSLayoutConstraint *leadingButtonConstraintTrailingWithTrailingButton;
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *trailingButtonConstraintBottom;
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *trailingButtonConstraintTop;
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *trailingButtonConstraintTrailing;
+
+@end
+
+@implementation MDCBannerView
+
+- (instancetype)initWithFrame:(CGRect)frame {
+  self = [super initWithFrame:frame];
+  if (self) {
+    [self commonBannerViewInit];
+  }
+  return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  self = [super initWithCoder:aDecoder];
+  if (self) {
+    [self commonBannerViewInit];
+  }
+  return self;
+}
+
+- (void)commonBannerViewInit {
+  self.backgroundColor = UIColor.whiteColor;
+
+  // Create textLabel
+  UILabel *textLabel = [[UILabel alloc] init];
+  textLabel.translatesAutoresizingMaskIntoConstraints = NO;
+  textLabel.font = [MDCTypography body2Font];
+  textLabel.textColor = UIColor.blackColor;
+  textLabel.alpha = [MDCTypography body2FontOpacity];
+  textLabel.numberOfLines = kTextNumberOfLineLimit;
+  textLabel.lineBreakMode = NSLineBreakByTruncatingTail;
+  [self addSubview:textLabel];
+  _textLabel = textLabel;
+
+  // Create imageView
+  UIImageView *imageView = [[UIImageView alloc] init];
+  imageView.translatesAutoresizingMaskIntoConstraints = NO;
+  if (@available(iOS 9.0, *)) {
+    [imageView.widthAnchor constraintEqualToConstant:kImageViewSideLength].active = YES;
+    [imageView.heightAnchor constraintEqualToConstant:kImageViewSideLength].active = YES;
+  }
+  imageView.contentMode = UIViewContentModeCenter;
+  imageView.clipsToBounds = YES;
+  imageView.hidden = YES;
+  [self addSubview:imageView];
+  _imageView = imageView;
+
+  // Create a button container to organize buttons
+  UIView *buttonContainerView = [[UIView alloc] init];
+  buttonContainerView.translatesAutoresizingMaskIntoConstraints = NO;
+  [self addSubview:buttonContainerView];
+  self.buttonContainerView = buttonContainerView;
+
+  // Create leadingButton and trailingButton
+  MDCButton *leadingButton = [[MDCButton alloc] init];
+  leadingButton.translatesAutoresizingMaskIntoConstraints = NO;
+  leadingButton.backgroundColor = UIColor.whiteColor;
+  [buttonContainerView addSubview:leadingButton];
+  _leadingButton = leadingButton;
+
+  MDCButton *trailingButton = [[MDCButton alloc] init];
+  trailingButton.translatesAutoresizingMaskIntoConstraints = NO;
+  trailingButton.backgroundColor = UIColor.whiteColor;
+  [buttonContainerView addSubview:trailingButton];
+  _trailingButton = trailingButton;
+
+  [self addObserversToSubviews];
+  if (@available(iOS 9_0, *)) {
+    [self setupConstraints];
+  }
+}
+
+#pragma mark - Constraints Helpers
+
+- (void)setupConstraints NS_AVAILABLE_IOS(9_0) {
+  self.imageViewConstraintLeading =
+  [self.imageView.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor
```

Comment:
Please use if(@available(iOS 9.0, *)){ here to protect against running it on iOS 8.

Reply:
Sure. I can do it. One thing I am thinking of is would https://github.com/material-components/material-components-ios/pull/7065/files/aeeecd3982520bfa91577f0db37cb0e67005d986#diff-98d48786cc5ed35541325aca00c9c007R155 be enough to guard it? Add another `@available` will be another layer of runtime of safeguard but does it mean all methods annotated with `NS_AVAILABLE_IOS` need to have this additional runtime safeguard internally? I am not sure if UIKit does it or not.

Reply:
If there's an alternative, then I would suggest that the solution above (version-checking before each invocation) is good. However, if there's no alternative, why generate so many `if` blocks? If it's always a no-op, make the method itself a no-op.  It's also much safer that way.

Reply:
I am confused. 

> If there's an alternative, then I would suggest that the solution above (version-checking before each invocation) is good.

What is the alternative for? Right now the version checking is placed when the method annotated `NS_AVAILABLE_IOS` everything (and it has to be placed to clam down warnings). From what you described, it is sufficient enough right? Please feel free to correct me if I understood it wrong.

Reply:
Addressed in 88e43257d5d16bda0726dac9632fddb2beb9290c.